Content.Client/GameObjects/EntitySystems/ContextMenuPresenter.cs

```diff
@@ -0,0 +1,309 @@
+﻿#nullable enable
+using System;
+using System.Linq;
+using System.Threading;
+using Content.Client.State;
+using Content.Client.UserInterface;
+using Content.Client.UserInterface.ContextMenu;
+using Content.Client.Utility;
+using Content.Shared;
+using Content.Shared.GameObjects.Verbs;
+using Content.Shared.Input;
+using Robust.Client.GameObjects;
+using Robust.Client.Input;
+using Robust.Client.Player;
+using Robust.Client.State;
+using Robust.Client.UserInterface;
+using Robust.Shared.Configuration;
+using Robust.Shared.Containers;
+using Robust.Shared.GameObjects;
+using Robust.Shared.Input;
+using Robust.Shared.Input.Binding;
+using Robust.Shared.IoC;
+using Robust.Shared.Map;
+using Robust.Shared.Timing;
+using Timer = Robust.Shared.Timers.Timer;
+namespace Content.Client.GameObjects.EntitySystems
+{
+    public class ContextMenuPresenter : IDisposable
+    {
+        [Dependency] private readonly IEntitySystemManager _systemManager = default!;
+        [Dependency] private readonly IItemSlotManager _itemSlotManager = default!;
+        [Dependency] private readonly IEntityManager _entityManager = default!;
+        [Dependency] private readonly IPlayerManager _playerManager = default!;
+        [Dependency] private readonly IStateManager _stateManager = default!;
+        [Dependency] private readonly IInputManager _inputManager = default!;
+        [Dependency] private readonly IConfigurationManager _cfg = default!;
+        [Dependency] private readonly IGameTiming _gameTiming = default!;
+
+        private readonly IContextMenuView _contextMenuView;
+        private readonly VerbSystem _verbSystem;
+
+        private bool _playerCanSeeThroughContainers;
+
+        private MapCoordinates _mapCoordinates;
+        private CancellationTokenSource? _cancellationTokenSource;
+
+        public ContextMenuPresenter(VerbSystem verbSystem)
+        {
+            IoCManager.InjectDependencies(this);
+            _verbSystem = verbSystem;
+            _verbSystem.ToggleContextMenu += SystemOnToggleContextMenu;
+            _verbSystem.ToggleContainerVisibility += SystemOnToggleContainerVisibility;
+
+            _contextMenuView = new ContextMenuView();
+            _contextMenuView.OnKeyBindDownSingle += OnKeyBindDownSingle;
+            _contextMenuView.OnMouseEnteredSingle += OnMouseEnteredSingle;
+            _contextMenuView.OnMouseExitedSingle += OnMouseExitedSingle;
+            _contextMenuView.OnMouseHoveringSingle += OnMouseHoveringSingle;
+
+            _contextMenuView.OnKeyBindDownStack += OnKeyBindDownStack;
+            _contextMenuView.OnMouseEnteredStack += OnMouseEnteredStack;
+
+            _contextMenuView.OnExitedTree += OnExitedTree;
+            _contextMenuView.OnCloseRootMenu += OnCloseRootMenu;
+            _contextMenuView.OnCloseChildMenu += OnCloseChildMenu;
+
+            _cfg.OnValueChanged(CCVars.ContextMenuGroupingType, _contextMenuView.OnGroupingContextMenuChanged, true);
```

Comment:
unsubscribe these events when disposing.

this is a nitpick but please add an empty line each between the methods of this file.

Reply:
Should probably be using the eventbus.

At the very least these need disposing.